### PR TITLE
Qt desktop GUI: upgrade to Qt6

### PR DESCRIPTION
**Pull request by SomberNight**
This PR migrates the Qt Widgets GUI used on desktop from Qt5 to Qt6.

Notes:
- the latest debian stable and ubuntu lts now both package python3-pyqt6, meaning most distros should now have it
- raises the minimum Windows version needed (for users) to win10 x64 (from 8.1 32 bit?)
  - see https://github.com/spesmilo/electrum/issues/6598
  - I think this is OK.
- raises the min macOS version needed to 11 (from 10.13)
  - related https://github.com/spesmilo/electrum/issues/3685
  - 10.13 was released in 2017-09, and 11 released in 2020-11. That's a significant bump, and the new minimum is only ~4 years old. Is this OK?
    - I think we bumped to 10.13 (from 10.11) around https://github.com/spesmilo/electrum/commit/095464b620e3e1bb699d174d969526f78ad356b9, which was released in 4.0.1 in [2020-07](https://github.com/spesmilo/electrum/blob/0f6898ed90be7777254b6868e1e1b0b5af79372f/RELEASE-NOTES#L1), at which point 10.13 was ~3 years old. Hence, the current bump is less strict and I think is OK.
- the win/mac/appimage binaries grew by ~20 MB :/
  - qt6 takes up more space. I tried to strip out some unneeded new modules (and adjusted our existing stripping-out to layout changes) - without that the size diff would be even greater.
- when running the AppImage on a system that has both wayland and xwayland, AFAICT previously QT_QPA_PLATFORM=xcb was the default, but now it is QT_QPA_PLATFORM=wayland
  - compare running `$ QT_QPA_PLATFORM=xcb ./run_electrum` vs `$ QT_QPA_PLATFORM=wayland ./run_electrum`
  - note that they look a bit different
	- in particular the titlebar for e.g. QMessageBox is missing on wayland
	  - e.g. try in console tab: `>>> window.question(msg="msg")`

closes https://github.com/spesmilo/electrum/issues/8007

### CODE REVIEW

File: electrum/gui/qt/qrreader/__init__.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 #   to access the camera
 # - zbar fails to access the camera on macOS
 # - qtmultimedia seems to support more cameras on Windows than zbar
-# - qtmultimedia is often not packaged with PyQt5
+# - qtmultimedia is often not packaged with PyQt
 #   in particular, on debian, you need both "python3-pyqt5" and "python3-pyqt5.qtmultimedia"
```
Comment (accumulator) [inline electrum/gui/qt/qrreader/__init__.py:15]:
```suggestion
#   in particular, on debian, you need both "python3-pyqt6" and "python3-pyqt6.qtmultimedia"
```

**Comment by accumulator** [comment]
Even though it seems to work, `QDarkStyle` uses `load_stylesheet_pyqt5()`.

https://github.com/SomberNight/electrum/blob/c3a869d5e1dd0d0884e333a44694eb14b16c7a4f/electrum/gui/qt/__init__.py#L179

There's a `load_stylesheet_pyqt6()` func since `qdarkstyle>=3.2`

https://github.com/SomberNight/electrum/blob/c3a869d5e1dd0d0884e333a44694eb14b16c7a4f/contrib/requirements/requirements.txt#L3

**Review by accumulator — CHANGES_REQUESTED** [review]

File: electrum/gui/common_qt/__init__.py [hunk 2]
```diff
@@ -10,7 +10,7 @@ def get_qt_major_version() -> int:
     _GUI_QT_VERSION = getattr(sys, '_GUI_QT_VERSION', None)
     if _GUI_QT_VERSION is None:
         # used by pyinstaller when building (analysis phase)
-        _GUI_QT_VERSION = 5
+        _GUI_QT_VERSION = 6
```
[the remark this answers is not in the document]
Comment (SomberNight) [reply electrum/gui/common_qt/__init__.py]: No reason to keep it. Removed now.

**Comment by SomberNight** [comment]
> There's a load_stylesheet_pyqt6() func since qdarkstyle>=3.2

Thanks, done.

----

At a higher level, what do you think about doing this migration now? :)  AFAICS there are no real blockers.

**Comment by accumulator** [comment]
> > There's a load_stylesheet_pyqt6() func since qdarkstyle>=3.2
> 
> Thanks, done.
> 
> At a higher level, what do you think about doing this migration now? :) AFAICS there are no real blockers.

It looks fully functional. The only thing I ran into testing the branch was an error when grabbing QR from screen. That was on both wayland and xcb backends.

File: contrib/build-linux/appimage/make_appimage.sh
Comment (SomberNight) [inline contrib/build-linux/appimage/make_appimage.sh:1]: > when running the AppImage on a system that has both wayland and xwayland, AFAICT previously QT_QPA_PLATFORM=xcb was the default, but now it is QT_QPA_PLATFORM=wayland 

> I didn't test the AppImage yet. There might be dependency issues between pyqt6 and debian buster.

I've now tried running the appimage on debian 10 (oldoldstable), and am getting an error with wayland.

```
  4.16 | D | util.profiler | Plugins.__init__ 0.0422 sec
  4.16 | I | daemon.Daemon | launching GUI: qt
  4.76 | I | gui.qt.ElectrumGui | Qt GUI starting up... Qt=6.7.1, PyQt=6.7.1
/tmp/.mount_electrFlGFOt/usr/bin/python3: symbol lookup error: /tmp/.mount_electrFlGFOt/usr/lib/python3.11/site-packages/PyQt6/Qt6/plugins/platforms/../../lib/libQt6WaylandClient.so.6: undefined symbol: wl_proxy_marshal_flags
```

If I explicitly specify `QT_QPA_PLATFORM=xcb`, it starts and works as expected. But it picks wayland by default.
I found https://bugreports.qt.io/browse/QTBUG-114635 and it looks like even debian 11 might be affected.

One commenter recommends setting the env var to prefer x11 at runtime:
```
import os
if sys.platform == "linux":
    os.environ.setdefault("QT_QPA_PLATFORM", "xcb")
```
What do you think of that?

Or should we perhaps try to package a newer libwayland into the appimage? I am not sure if that is realistic.

Comment (accumulator) [reply contrib/build-linux/appimage/make_appimage.sh:1]: Maybe just document this, and let users of oldoldstable era distro's set the `xcb` override. This way we don't have to nag users on more recent distro's who probably expect it to properly detect the window system.

Comment (SomberNight) [reply contrib/build-linux/appimage/make_appimage.sh:1]: Well based on that thread, what is needed is `libwayland-client0>=1.19`, which is too old even for debian 11 (oldstable), or for ubuntu 20.04 lts. Meaning users would need current debian stable or ubuntu 22+.

What if we scoped it to `if sys.platform == "linux" and os.environ.get("APPIMAGE")`?

Ideally we could tell Qt to try the xcb plugin as a fallback...

Comment (accumulator) [reply contrib/build-linux/appimage/make_appimage.sh:1]: > What if we scoped it to `if sys.platform == "linux" and os.environ.get("APPIMAGE")`?

Yeah, better to only do this for AppImage builds

Comment (SomberNight) [reply contrib/build-linux/appimage/make_appimage.sh:1]: done in https://github.com/spesmilo/electrum/pull/9189/commits/0fb20cc816f83037664e08d701e260a13fb50785

**Comment by SomberNight** [comment]
> The only thing I ran into testing the branch was an error when grabbing QR from screen. That was on both wayland and xcb backends.

I think that functionality is already broken on Linux if wayland is being used as window manager (regardless of the qpa plugin used). It might still work if wayland is not used at all.
On Windows and macOS, it works though.

**Pull request merged**